keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,727 @@
+
+### New Kubelet Configuration
+
+We will be introducing two new parameters to `KubeletConfiguration struct`
+defined in
+[https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go).
+These two configurations, if set, will apply to every container of the Node
+where kubelet is running.
+
+|Name|Description|Default Value|Feature Gate|
+|--- |--- |--- |--- |
+|MemorySwapLimit|This parameter sets total memory limit (memory + swap). This limits the total amount of memory this container is allowed to swap to disk.|-2, which enable disable swap|SupportNodeMemorySwap|
+|MemorySwappiness|This configuration sets how aggressively the kernel will swap memory pages. By default, the host kernel can swap out a percentage of anonymous pages used by a container. Users can set value between 0 and 100, to tune this percentage.|Unset, which will use host value|SupportNodeMemorySwap|
+
+#### MemorySwapLimit details
+
+MemorySwapLimit configuration is a kubelet flag that only takes effect on a
+container that has a memory limit set, either explicitly from
+[PodSpec]([https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits)
+) or implicitly from [Resource
+Quota]([https://kubernetes.io/docs/concepts/policy/resource-quotas/](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+).
+
+For container with memory limit set, MemorySwapLimit setting will have the
+following effects, [similar to
+docker](https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details)
+
+* If MemorySwapLimit is set to a positive integer,
+  * If the memory limit of the container is greater or equal to
+    MemorySwapLimit, then no swap is allowed, the container does not have
+    access to swap.
+  * If the memory limit of the container is less than MemorySwapLimit, then
+    MemorySwapLimit represents the total amount of memory and swap that can be
+    used. For example, for a container with memory limit set to 300m, and
+    `MemorySwapLimit` set to 1g, the container can use 300m of memory and 700m (1g
+    - 300m) swap.
+* If MemorySwapLimit is set to 0, for containers with memory limit is set, the
+  container can use as much swap as the Memory limit setting, if the host
+  container has swap memory configured. For instance, if  a container requests
+  memory="300m" and MemorySwapLimit is not set, the container can use 600m in
+  total of memory and swap.
+* If MemorySwapLimit is explicitly set to -1, the container is allowed to use
+  unlimited swap, up to the amount available on the host system.
+* If MemorySwapLimit is explicitly set to -2,  the container does not have
+  access to swap. This value effectively prevents a container from using swap.
```

Comment:
I know these config flags are intended to follow the docker conventions, but this use of 0/-1/-2 here is non-standard.

Could we make this an int pointer, where, if it is not set, we default to current behaviour/no swap usage, and if it is set, 0 and -1 follow the docker conventions?

Reply:
Yeah, if 0/-1/-2 is non-standard K8s convention, then we'll update to comply. 

How about string instead? so that we can pass in things like  "1G", "1Mi" as well, and make some of the intention explicit. Specifically, we want to introduce an easy way to say we do not allow container to access any swap. It will be a useful default value when `fail-swap-on` flag value is flipped. 

* -1 => "unlimited"
* -2 => "no-swap"
* 0/unset => ""
* X => "Xu" eg: "1g", "1mi", etc

Reply:
Standard convention IME is int pointer when you want different behaviour for set vs. unset. We can ask for some API review feedback. I don't think we should use a string since we'll have to convert to int in the CRI.

Reply:
SGTM, we can rely on API review feedback on this part and go with the recommendation. Regardless of the type specifics, I think the key point is to introduce an explicit value for disabling all swap use across container as a safety latch.